### PR TITLE
Move column stats collection to per-column rather than per-row logic

### PR DESCRIPTION
**Pull request by romseygeek**
Refactors the column stats collection logic so that each column is handled
by a separate ColumnStatsBuilder.  This will allow us to specialise how different
column types build their stats.

Relates to #15152

### CODE REVIEW

**Comment by romseygeek** [comment]
Note that this is a PR against a feature branch, `feature/analyze-datasketches`.  As it stands it will probably be less performant than the existing implementation as sample merging is now done separately for each column rather than once for all samples, but follow-ups will change the ColumnStatsBuilder implementation to use data sketches rather than holding samples as a `List<T>` so should dramatically reduce the amount of RAM required to build the stats.

File: .idea/inspectionProfiles/Project_Default.xml [hunk 1]
```diff
@@ -1,10 +1,14 @@
...
   <profile version="1.0">
     <option name="myName" value="Project Default" />
+    <inspection_tool class="AutoCloseableResource" enabled="true" level="WARNING" enabled_by_default="true">
+      <option name="METHOD_MATCHER_CONFIG" value="java.util.Formatter,format,java.io.Writer,append,com.google.common.base.Preconditions,checkNotNull,org.hibernate.Session,close,java.io.PrintWriter,printf,java.io.PrintStream,printf,org.apache.lucene.index.LeafReaderContext,reader" />
+    </inspection_tool>
     <inspection_tool class="ForCanBeForeach" enabled="false" level="WARNING" enabled_by_default="false">
       <option name="REPORT_INDEXED_LOOP" value="true" />
       <option name="ignoreUntypedCollections" value="false" />
     </inspection_tool>
+    <inspection_tool class="NewClassNamingConvention" enabled="false" level="WARNING" enabled_by_default="false" />
```
Comment (mfussenegger) [inline .idea/inspectionProfiles/Project_Default.xml:11]: Are these changes intentional? If so - can you move them into a dedicated PR that explains the change?

File: server/src/main/java/io/crate/statistics/ColumnStatsBuilder.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+            streamer.writeValueTo(out, samples.get(i));
+        }
+    }
+
+    public void add(T value, RamAccounting ramAccounting) {
+        sampleCount++;
+        if (value == null) {
+            nullCount++;
+        } else {
+            totalBytes += dataType.valueBytes(value) + Long.BYTES;
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/statistics/ColumnStatsBuilder.java:78]: Can you add a short comment explaining the extra `+ Long.BYTES` ?

File: server/src/main/java/io/crate/execution/engine/fetch/FetchIdIterator.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import java.io.IOException;
+
+import org.apache.lucene.index.IndexReaderContext;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.ReaderUtil;
+
+import com.carrotsearch.hppc.LongArrayList;
+import com.carrotsearch.hppc.cursors.LongCursor;
+
+public abstract class FetchIdIterator {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/engine/fetch/FetchIdIterator.java:33]: Is there a reason for the abstract class? There's currently only a single implementation which could be used directly?

File: server/src/main/java/io/crate/statistics/ColumnStatsBuilder.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+import java.util.Random;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import io.crate.Streamer;
+import io.crate.data.breaker.RamAccounting;
+import io.crate.types.DataType;
+
+public class ColumnStatsBuilder<T> {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/statistics/ColumnStatsBuilder.java:37]: Will this later contain the sketches? Currently it seems strange that this is streamable too, instead of streaming the result directly.

Maybe worth adding a unit test for this. There are some codecov warnings about missing coverage for the `merge`

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java [hunk 5]
```diff
@@ -183,159 +159,211 @@ Samples getSamples(RelationName relationName, List<Reference> columns, int maxSa
...
+            assert collectors.size() == expressionsAndInputs.expressions.size();
+            assert collectors.size() == expressionsAndInputs.inputs.size();
+            for (int i = 0; i < collectors.size(); i++) {
+                collectors.get(i).setShard(expressionsAndInputs.expressions.get(i), expressionsAndInputs.inputs.get(i));
             }
-            records.add(new RowN(row));
         }
 
-        return records;
     }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/statistics/ReservoirSampler.java:311]: Would it make sense to re-use `ExpressionsInput` here? Part of this looks similar.

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java [hunk 6]
```diff
@@ -183,159 +159,211 @@ Samples getSamples(RelationName relationName, List<Reference> columns, int maxSa
...
+        ReaderContext currentLeafContext;
+        int rowsCollected;
 
-        DocIdToRow(Engine.Searcher searcher,
-                   List<Input<?>> inputs,
-                   List<? extends LuceneCollectorExpression<?>> expressions) {
-            this.searcher = searcher;
-            this.inputs = inputs;
+        private ColumnSampler(
+            List<ColumnCollector<?>> expressions,
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/statistics/ReservoirSampler.java:324]:
```suggestion
            List<ColumnCollector<?>> collectors,
```

It's a bit confusing to use `expressions` here, which we usually use for instances of `CollectExpression`. And there's also `ShardExpressions` here, together with the use of `var` - makes it difficult to keep track of what's what if everything is expressions.

File: server/src/main/java/io/crate/statistics/ColumnStatsBuilder.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        for (int i = 0; i < maxSampleCount; i++) {
+            mergedStats.add(sampler.next(), RamAccounting.NO_ACCOUNTING);
+        }
+        mergedStats.nullCount = this.nullCount + other.nullCount;
+        mergedStats.sampleCount = this.sampleCount + other.sampleCount;
+        return mergedStats;
+    }
+
+    // https://ballsandbins.wordpress.com/2014/04/13/distributedparallel-reservoir-sampling/
+    private static class Sampler<T> {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/statistics/ColumnStatsBuilder.java:126]: Any reason to extract this into a class instead of keeping it inline as before? The state is available as locals in the one spot it's used, and it has just a single method.

File: .idea/inspectionProfiles/Project_Default.xml (hunk 1 above)
Comment (romseygeek) [reply .idea/inspectionProfiles/Project_Default.xml:11]: Ah, no, that shouldn't be there.  Although they're probably worth adding to the main project as they silence some pointlessly noisy warnings.  Will open a separate PR.

File: server/src/main/java/io/crate/execution/engine/fetch/FetchIdIterator.java (hunk 3 above)
Comment (romseygeek) [reply server/src/main/java/io/crate/execution/engine/fetch/FetchIdIterator.java:33]: It's the sort of thing that I've found generally useful in the past, but you're right, it doesn't need to be an abstract class yet.  Will merge it into the implementation.

File: server/src/main/java/io/crate/statistics/ColumnStatsBuilder.java (hunk 4 above)
Comment (romseygeek) [reply server/src/main/java/io/crate/statistics/ColumnStatsBuilder.java:37]: The followup will actually break this into two parts: the Builder which does the sampling, and then a separate immutable `ColumnSketch<T>` class which will handle serialization and merging.  Given that I'm not sure it's worth adding unit tests yet.

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java (hunk 5 above)
Comment (romseygeek) [reply server/src/main/java/io/crate/statistics/ReservoirSampler.java:311]: This is a great idea, I'll see if I can simplify things out by using the lucene expressions directly.

File: server/src/main/java/io/crate/statistics/ColumnStatsBuilder.java (hunk 4 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/statistics/ColumnStatsBuilder.java:37]: Okay, fine to keep as is if it will change anyway before merge to master.

File: server/src/main/java/io/crate/statistics/ColumnStatsBuilder.java (hunk 7 above)
Comment (romseygeek) [reply server/src/main/java/io/crate/statistics/ColumnStatsBuilder.java:126]: Yeah that can stay inline, I think this started off being more complicated and then got simpler again without me re-inlining it.

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java (hunk 5 above)
Comment (romseygeek) [reply server/src/main/java/io/crate/statistics/ReservoirSampler.java]: This has simplified things a lot.  I think we will be able to simplify further when we split the ColumnStatsBuilder into a local ColumnSketchBuilder and a serializable ColumnSketch, as we can then make the builder directly extend LuceneCollectorExpression and remove the ColumnSampler class.

File: server/src/main/java/io/crate/statistics/ReservoirSampler.java [hunk 8]
```diff
@@ -183,159 +161,233 @@ Samples getSamples(RelationName relationName, List<Reference> columns, int maxSa
...
+
+            return new Samples(statsBuilders, totalNumDocs, totalSizeInBytes);
+        } finally {
+            for (var shard : searchersToRelease) {
+                shard.searcher.close();
+            }
+        }
+    }
+
+    private List<? extends LuceneCollectorExpression<?>> getCollectorExpressions(
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/statistics/ReservoirSampler.java:250]:
```suggestion
    private static List<? extends LuceneCollectorExpression<?>> getCollectorExpressions(
```
